packages/tasit-action/src/TasitAction.js

```diff
@@ -19,117 +22,370 @@ class Utils {
+      this.removeListener(eventName, listener);
+    }, EVENTS_CONFIG.timeout);
+  };
+}
+
+class TransactionSubscription extends Subscription {
   #txPromise;
   #provider;
+  #tx;
 
   constructor(txPromise, provider) {
+    super(provider);
     this.#txPromise = txPromise;
     this.#provider = provider;
   }
 
-  on = async (trigger, callback) => {
-    if (trigger !== "confirmation") {
-      throw new Error(`Invalid subscription trigger, use: ["confirmation"]`);
+  on = async (eventName, listener) => {
+    const triggers = ["confirmation", "error"];
+
+    if (!triggers.includes(eventName)) {
+      throw new Error(`Invalid listener trigger, use: [${triggers}]`);
     }
 
-    if (!callback || typeof callback !== "function") {
-      throw new Error(`Cannot subscribe without a function`);
+    if (!listener || typeof listener !== "function") {
+      throw new Error(`Cannot listening without a function`);
     }
 
-    const blockHeight = 1;
-    const tx = await this.#txPromise;
+    if (eventName === "error") {
+      this.addErrorListener(listener);
+      return;
+    }
 
-    this.#provider.on(tx.hash, async receipt => {
-      const { confirmations } = receipt;
-      const message = {
-        data: {
-          confirmations: confirmations,
-        },
-      };
+    this.#tx = await this.#txPromise;
 
+    let wrappedListener = async blockNumber => {
       try {
-        await callback(message);
+        const receipt = await this.#provider.getTransactionReceipt(
+          this.#tx.hash
+        );
+        const txConfirmed = receipt !== null;
+
+        if (txConfirmed) {
+          this.setEventInfo("block", wrappedListener, {
+            hasBeenConfirmed: true,
+          });
+        } else {
+          const { hasBeenConfirmed } = this.getEventInfo(
+            "block",
+            wrappedListener
+          );
+          if (hasBeenConfirmed)
+            this.emitErrorEvent(
+              new Error(`Your message has been included in an uncle block.`)
+            );
+
+          return;
+        }
+
+        const { confirmations } = receipt;
+        const message = {
+          data: {
+            confirmations: confirmations,
+          },
+        };
+
+        await listener(message);
       } catch (error) {
-        throw new Error(`Callback function with error: ${error.message}`);
+        this.emitErrorEvent(
+          new Error(`Callback function with error: ${error.message}`)
+        );
       }
-    });
+    };
+
+    this.addListener(eventName, "block", listener, wrappedListener);
   };
 
-  removeAllListeners = async () => {
+  // Tech debt
+  // This method avoid duplicated nonce generation when rapid succession of several transactions
+  // See: https://github.com/ethereumbook/ethereumbook/blob/04f66ae45cd9405cce04a088556144be11979699/06transactions.asciidoc#keeping-track-of-nonces
+  // How we'll should keeping track of nonces?
+  waitForMessage = async () => {
```

Comment:
I think that likely will be used on SDK. Nonce tracking isn't easy to do.
Without `waitForMessage` users should nest messages after confirmation inside callbacks.

Reply:
Wait, what is the "message" it's waiting for again? If it's one confirmation, even that's not necessarily enough time to wait. Unless the message we're waiting for is just a new nonce, in which case that would be enough.

Reply:
I guess this is two questions:

1) By message, do you mean a general term for transactions and meta-transactions? If so, I think "message" is two broad since messages aren't always destined to create transactions.

2) Is there data we're waiting for in response to the write before we can move on? Or just the knowledge that the write has happened "successfully"?

Reply:
I'm trying to use `transaction` word on API. To keep SDK abstraction level idea. So, until now I'm using `message` word to express `contract changing state transaction`:
`normal-tx` and `meta-tx` (messages) are broadcasted (direct or indirect) to blockchain and their success or failing (response) will be caught by an event.
As meta-tx is a missing part yet, I'm not sure if that thought/abstraction is correct.
The technical means of the `waitForMessage` maybe are: `waitForMessageBeReceivedByBlockchain`, `waitForMessageBeBroadcast`,
or even `waitForGetTransactionCountGetRightValue` (Obs: there aren't function name sugestions).

> When you build an application that constructs transactions, it cannot rely on getTransactionCount for pending transactions. Only when the pending and confirmed counts are equal (all outstanding transactions are confirmed) can you trust the output of getTransactionCount to start your nonce counter. Thereafter, keep track of the nonce in your application until each transaction confirms.

Refs: https://github.com/ethereumbook/ethereumbook/blob/04f66ae45cd9405cce04a088556144be11979699/06transactions.asciidoc#keeping-track-of-nonces

Reply:
On which API? The Tasit SDK API? Based on what you typed after that first sentence, did you mean that you're trying _not_ to use the transaction word on the API?

I still think that message is too generic, because it might just be an Ethereum whisper message that has no "ambitions" to eventually become a transaction.

But I like the goal of finding a word to express `contract-changing state (hopefully eventual) transactions`. Maybe `intendedTransaction`? Ooh, or what about `action`? That way it has the same name as `tasit-action`.

I think the abstraction you're picturing is correct.

I suppose all of those functions (not with those names, but that functionality) might be needed eventually. In particular, it sounds like `waitForMessageToBeRecievedByBlockchain` isn't enough until it's confirmed. Same with `waitForMessageToBeBroadcast`. The thing this function tries to do is `waitForTransactionConfirmation` (I don't think meta-transactions get confirmed and thus update the nonce automatically? If so we need to address that.)

I'm all for naming functions based on their motivations over the things they do. So maybe we call this function `waitForNonceToUpdate`? And add a comment in the code to note that what needs to happen for the nonce to update for a meta-transaction is an open question?

Reply:
> On which API? The Tasit SDK API? Based on what you typed after that first sentence, did you mean that you're trying not to use the transaction word on the API?

Sorry! NOT to use `transaction` word!

> Ooh, or what about action? That way it has the same name as tasit-action.

Yeah, `action` seems perfect!

> The thing this function tries to do is waitForTransactionConfirmation (I don't think meta-transactions get confirmed and thus update the nonce automatically? If so we need to address that.)

I'll add that on [meta-tx issue](https://github.com/tasitlabs/TasitSDK/issues/88), I'll look closer how that works.

> I'm all for naming functions based on their motivations over the things they do. So maybe we call this function waitForNonceToUpdate? And add a comment in the code to note that what needs to happen for the nonce to update for a meta-transaction is an open question?

Could be more abstract name better? As avoiding `transaction` word, Isn't `nonce` a too low-level / blockchain related word?

Reply:
I'll update code to use `action` insted `message` word.
A question: Which is better: use that only to exposed methods or name TasitAction implementation as well (E.g.: `TransactionSubscription` => `ActionSubscription`)?

Reply:
> Sorry! NOT to use `transaction` word!

Understood. Thanks for clarifying.

> Yeah, `action` seems perfect!

Glad you like the word `action`.

> I'll add that on meta-tx issue, I'll look closer how that works.

I pasted the body of that text on the meta-tx issue too.

> Could be more abstract name better? As avoiding `transaction` word, Isn't `nonce` a too low-level / blockchain related word?

Yes, I agree. But the part I'm finding tricky is, I can't find a non-technical way to express why we're waiting here. Because if we just call it waitForAction, as a developer using this SDK I would go back to the question I first had here - why even wait? The only answer is a technical one. Maybe instead of nonce we could say `waitForUpdatedActionId` or something. Because this is a little confusing, we should add a safety check wherever possible to make sure user waits where they should by setting a boolean and then checking that. We can create an issue for this, perhaps.

> I'll update code to use `action` insted `message` word.
A question: Which is better: use that only to exposed methods or name TasitAction implementation as well (E.g.: `TransactionSubscription` => `ActionSubscription`)?

Great! Hmm, great question. I'm leaning towards renaming to `ActionSubscription` if we'll. use it for meta-tx too eventually, but `TransactionSubscription` if we won't.

Reply:
Ok,
- I'll change `message` & `transaction` to `action` whenever possible.
- I'll rename `waitForMessage` to `waitForNonceToUpdate` for now and add a note comment.
- Meta-tx open questions will be care on https://github.com/tasitlabs/TasitSDK/issues/88

Is it ok?

Reply:
Sounds great!